### ISSUE TITLE
Usage of prefix with LogstashEncoder and LayoutWrappingEncoder throws errors

### ISSUE THREAD

**Issue by nhakimjee** [logstash/logstash-logback-encoder#232]
I am trying to send my application logs to syslog using LogstashTcpSocketAppender. My configuration is as follows:

<appender name="stash-tcp" class="net.logstash.logback.appender.LogstashTcpSocketAppender">
      
      <destination>127.0.0.1:514</destination>
      <reconnectionDelay>30 second</reconnectionDelay>
      <keepAliveDuration>5 minutes</keepAliveDuration>
      <writeBufferSize>16384</writeBufferSize>

      
      <encoder class="net.logstash.logback.encoder.LogstashEncoder">
      	 <prefix class="ch.qos.logback.core.encoder.LayoutWrappingEncoder">
            <layout class="ch.qos.logback.classic.PatternLayout">
               <pattern>%syslogStart{LOCAL0}</pattern>
            </layout>
         </prefix>

        <fieldNames>
    	  <thread>[ignore]</thread>
    	  <levelValue>[ignore]</levelValue>
	</fieldNames>

        <includeCallerData>true</includeCallerData>      
      </encoder>
      
      <includeCallerData>true</includeCallerData>
 </appender>

However, I get the following errors when I startup my web application.
09:21:48,936 |-INFO in ch.qos.logback.classic.LoggerContext[default] - Found resource [logback-test.xml] at [file:/home/nhakimjee/work/Workspaces/syslog/.metadata/.plugins/org.eclipse.wst.server.core/tmp0/wtpwebapps/bap/WEB-INF/classes/logback-test.xml]
09:21:49,001 |-INFO in ch.qos.logback.classic.joran.action.ConfigurationAction - debug attribute not set
09:21:49,002 |-INFO in ch.qos.logback.core.joran.action.ConversionRuleAction - registering conversion word syslogStart with class [ch.qos.logback.classic.pattern.SyslogStartConverter]
09:21:49,002 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - About to instantiate appender of type [net.logstash.logback.appender.LogstashTcpSocketAppender]
09:21:49,028 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - Naming appender as [stash-tcp]
09:21:49,358 |-ERROR in ch.qos.logback.core.joran.util.PropertySetter@5475d1b2 - A "net.logstash.logback.encoder.LogstashEncoder" object is not assignable to a "ch.qos.logback.core.Appender" variable.
09:21:49,358 |-ERROR in ch.qos.logback.core.joran.util.PropertySetter@5475d1b2 - The class "ch.qos.logback.core.Appender" was loaded by 
09:21:49,359 |-ERROR in ch.qos.logback.core.joran.util.PropertySetter@5475d1b2 - [WebappClassLoader
  context: ROOT
  delegate: false
----------> Parent Classloader:
java.net.URLClassLoader@2f92e0f4
] whereas object of type 
09:21:49,359 |-ERROR in ch.qos.logback.core.joran.util.PropertySetter@5475d1b2 - "net.logstash.logback.encoder.LogstashEncoder" was loaded by [WebappClassLoader
  context: ROOT
  delegate: false
----------> Parent Classloader:
java.net.URLClassLoader@2f92e0f4
].
09:21:49,359 |-INFO in ch.qos.logback.core.joran.action.NestedComplexPropertyIA - Assuming default type [net.logstash.logback.fieldnames.LogstashFieldNames] for [fieldNames] property

I used the examples to for the above configuration, am I missing some something?

Thanks,

Nuruddin

**Comment by philsttr** [comment]
Hi Nuruddin,

Can you please provide your complete logback.xml file (while masking any sensitive data)?  Specifically, I need to see the appender element.

**Comment by nhakimjee** [comment]
Sorry for the delay. Here is my appender:

	<conversionRule conversionWord="syslogStart" converterClass="ch.qos.logback.classic.pattern.SyslogStartConverter"/>
	
	<appender name="stash-tcp" class="net.logstash.logback.appender.LogstashTcpSocketAppender">
	  
      <destination>127.0.0.1:514</destination>
      <reconnectionDelay>30 second</reconnectionDelay>
      <keepAliveDuration>5 minutes</keepAliveDuration>
	  <writeBufferSize>16384</writeBufferSize>

      
      <encoder class="net.logstash.logback.encoder.LogstashEncoder">
      	<prefix class="ch.qos.logback.core.encoder.LayoutWrappingEncoder">
          <layout class="ch.qos.logback.classic.PatternLayout">
            <pattern>%syslogStart{LOCAL6}</pattern>
          </layout>
        </prefix>

        <fieldNames>
    	  <thread>[ignore]</thread>
    	  <levelValue>[ignore]</levelValue>
	    </fieldNames>

        <includeCallerData>true</includeCallerData>      
      </encoder>
      
      <includeCallerData>true</includeCallerData>
  	</appender>

**Comment by philsttr** [comment]
Looks like this is caused by a backwards incompatible change introduced in logback 1.2.1.  Specifically, [this change](https://github.com/qos-ch/logback/commit/d3d704c12e7b8e6503187744ac2200c244088c63)

For now, you can workaround this by using logback prior to 1.2.1.

I tested your config using logback 1.2.0 and 1.1.7, and it worked in both cases.

**Comment by philsttr** [comment]
Submitted [LOGBACK-1326](https://jira.qos.ch/browse/LOGBACK-1326) with https://github.com/qos-ch/logback/pull/383 to hopefully fix this in logback.

**Comment by nhakimjee** [comment]
Thank you!

**Comment by imod** [comment]
this does not look to be resolved in any foreseeable time - is there any other way this could be fixed?

**Comment by philsttr** [comment]
According to [this comment](https://github.com/qos-ch/logback/pull/383#issuecomment-469370401),   @ceki has marked [LOGBACK-1326](https://jira.qos.ch/browse/LOGBACK-1326) to be fixed in logback 1.3.0-alpha5.  And it looks like he has been active recently in the logback repo, so hopefully, this will be addressed soon.  I know 2 years is a long time to wait for [a PR](https://github.com/qos-ch/logback/pull/383) to be merged, but I'm still hopeful due to the recent activity.

Without the requested change in logback, the only workaround would be to copy the `LayoutWrappingAppender` class from logback, and make the change from [the PR](https://github.com/qos-ch/logback/pull/383)  Specifically, change the type of the `parent` field and `setParent` method parameter to `ContextAware`.

**Comment by floruschbaschan** [comment]
While working with `logback-core` as well as `logback-classic` version 1.2.3 we also run into the exception described above (`"net.logstash.logback.encoder.LogstashEncoder" object is not assignable to a "ch.qos.logback.core.Appender"`). We noticed that the order of elements defined in your appender seems to matter. We managed to define our own log pattern for the `LogstashEncoder` by one of the following configurations:

```
<?xml version="1.0" encoding="UTF-8"?>
<configuration>
	<!-- Option 1 - Multiple <encoder> -->
	<appender name="CONSOLE" class="ch.qos.logback.core.ConsoleAppender">
		<encoder class="net.logstash.logback.encoder.LogstashEncoder"/>
		<encoder>
			<pattern>${OUR_OWN_LOG_PATTERN}</pattern>
		</encoder>
	</appender>

	<!-- Option 2 - Using <layout> -->
	<appender name="CONSOLE" class="ch.qos.logback.core.ConsoleAppender">
		<encoder class="net.logstash.logback.encoder.LogstashEncoder"/>
		<layout class="ch.qos.logback.classic.PatternLayout">
			<pattern>${OUR_OWN_LOG_PATTERN}</pattern>
		</layout>
	</appender>

	<!-- Option 3 - Using LoggingEventCompositeJsonEncoder -->
	<appender name="CONSOLE" class="ch.qos.logback.core.ConsoleAppender">
		<encoder class="net.logstash.logback.encoder.LoggingEventCompositeJsonEncoder">
			<providers>
				<timestamp/>
				<version/>
				<pattern>
					<pattern>
						{
						"logger": "%logger",
						"level": "%level",
						"thread": "%thread",
						"message": "%message",
						"traceId": "%mdc"
						}
					</pattern>
				</pattern>
			</providers>
		</encoder>
	</appender>
</configuration>
```

Defining the log pattern before the `LogstashEncoder`, results in the default log pattern. Therefore we would assume that the order is important here. These approaches would allow us to customize the log pattern definition without the need to downgrade any versions of the logback dependencies. 

However, all examples I have seen are using only one encoder and put `LogstashEncoder` right at the end before </appender>. We are not sure if there are any disadvantages for one of the options above. Any help/advice on this would be appreciated.

**Comment by philsttr** [comment]
[LOGBACK-1326](https://jira.qos.ch/browse/LOGBACK-1326) has been fixed in logback 1.2.5
